handlers/rsa.go

```diff
@@ -0,0 +1,211 @@
+nwIBAw==
+-----END PUBLIC KEY-----`,
+}
+
+// RsaHandler is a handler that allows only 1 registration for IP
+type RsaHandler struct {
+	kv       db.Database
+	keysLock sync.RWMutex
+}
+
+func (rh *RsaHandler) addKey(index, value []byte) {
+	rh.keysLock.Lock()
+	defer rh.keysLock.Unlock()
+	tx := rh.kv.WriteTx()
+	defer tx.Discard()
+	if err := tx.Set(index, value); err != nil {
+		log.Error(err)
+	}
+	if err := tx.Commit(); err != nil {
+		log.Error(err)
+	}
+}
+
+func (rh *RsaHandler) exist(index []byte) bool {
+	rh.keysLock.RLock()
+	defer rh.keysLock.RUnlock()
+	tx := rh.kv.WriteTx()
+	defer tx.Discard()
+	_, err := tx.Get(index)
+	return err == nil
+}
+
+// GetName returns the name of the handler
+func (rh *RsaHandler) GetName() string {
+	return "rsa"
+}
+
+// Init initializes the handler.
+// Takes one argument for persistent data directory.
+func (rh *RsaHandler) Init(opts ...string) (err error) {
+	rh.kv, err = metadb.New(db.TypePebble, filepath.Clean(opts[0]))
+	return err
+}
+
+// Auth is the handler for the rsa handler
+func (rh *RsaHandler) Auth(r *http.Request,
+	ca *csp.Message, pid []byte, st string) (bool, string) {
+	log.Infof(r.UserAgent())
+
+	voterId, message, signature, err := parseRsaAuthData(ca.AuthData)
+	if err != nil {
+		log.Error(err)
+		return false, err.Error()
+	}
+
+	rsaPublickey, err := parseRsaPublicKey()
+	if err != nil {
+		log.Error(err)
+		return false, err.Error()
+	}
+
+	// Verify signature
+	if err := validateRsaSignature(signature, message, rsaPublickey); err != nil {
+		log.Warnf("invalid signature: %s\n", err)
+		return false, "invalid signature"
+	}
+
+	if st == csp.SignatureTypeSharedKey {
+		return true, "please, do not share the key"
+	}
+	if rh.exist(voterId) {
+		log.Warnf("%s already registered", voterId)
+		return false, "already registered"
+	}
+
+	rh.addKey(voterId, nil)
+	log.Infof("new user registered with id %s", voterId)
+
+	return true, ""
+}
+
+// RequireCertificate must return true if the auth handler requires some kind of client
+// TLS certificate. If true then CertificateCheck() and HardcodedCertificate() methods
+// must be correctly implemented. Else both function can just return true and nil.
+func (rh *RsaHandler) RequireCertificate() bool {
+	return false
+}
+
+// CertificateCheck is used by the Auth handler to ensure a specific certificate is
+// added to the CA cert pool on the HTTP/TLS layer (optional).
+func (rh *RsaHandler) CertificateCheck(subject []byte) bool {
+	return true
+}
+
+// Certificates returns a hardcoded CA certificated that will be added to the
+// CA cert pool by the handler (optional).
+func (rh *RsaHandler) Certificates() [][]byte {
+	return nil
+}
+
+// Internal data handlers
+
+func parseRsaPublicKey() (*rsa.PublicKey, error) {
+	if len(rsaPubKeys) < 1 {
+		return nil, errors.Errorf("no public keys")
+	}
+	// Using the first one available so far
+	block, _ := pem.Decode([]byte(rsaPubKeys[0]))
```

Comment:
I'm not sure what the second variable is but if that's an error, let's not skip it please :)

Reply:
Absolutely. Indeed, the string is supposed to be provided by us, and not to be faulty, but obviously, the if statement below was useless.

Reply:
If the input that is not supposed to be provided faulty, then the error should be check and panic or log.Fatal() should be called